lib/csu/common-cheri/crt_init_globals.c

```diff
@@ -146,43 +146,50 @@ do_crt_init_globals(const Elf_Phdr * __capability phdr, long phnum)
 		readonly_start = text_start;
 		readonly_end = text_end;
 	}
-	rodata_cap = cheri_clearperm(phdr,
-	    CHERI_PERM_EXECUTE | CHERI_PERM_STORE | CHERI_PERM_STORE_CAP |
-	    CHERI_PERM_STORE_LOCAL_CAP);
-	rodata_cap = cheri_setaddress(rodata_cap, readonly_start);
-	rodata_cap =
-	    cheri_setbounds(rodata_cap, readonly_end - readonly_start);
-
-	void * __capability data_cap = NULL;
 	if (!have_data_segment) {
 		/*
 		 * There cannot be any capabilities to initialize if there
 		 * is no data segment. Set all to NULL to catch errors.
 		 *
 		 * Note: RELRO segment will be part of a R/W PT_LOAD.
 		 */
-		rodata_cap = NULL;
 		code_cap = NULL;
+		data_cap = NULL;
+		rodata_cap = NULL;
 	} else {
 		/* Check that ranges are well-formed */
 		if (writable_end < writable_start ||
 		    readonly_end < readonly_start ||
 		    text_end < text_start)
 			__builtin_trap();
+
 		/* Abort if text and writeable overlap: */
 		if (MAX(writable_start, text_start) <=
 		    MIN(writable_end, text_end)) {
 			/* TODO: should we allow a single RWX segment? */
 			__builtin_trap();
 		}
-		data_cap =
-		    cheri_setaddress(__DECONST_CAP(void * __capability, phdr),
-		    writable_start);
-		/* Bound the result and clear execute permissions. */
+
+#ifdef __CHERI_PURE_CAPABILITY__
+		data_cap = __DECONST_CAP(void * __capability, phdr);
+#else
+		data_cap = cheri_getdefault();
+#endif
 		data_cap = cheri_clearperm(data_cap, CHERI_PERM_EXECUTE);
+
+		code_cap = cheri_getpcc();
+		rodata_cap = cheri_clearperm(data_cap,
+		    CHERI_PERM_STORE | CHERI_PERM_STORE_CAP |
+		    CHERI_PERM_STORE_LOCAL_CAP);
+
+		data_cap = cheri_setaddress(data_cap, writable_start);
+		rodata_cap = cheri_setaddress(rodata_cap, readonly_start);
+
 		/* TODO: should we use exact setbounds? */
 		data_cap =
 		    cheri_setbounds(data_cap, writable_end - writable_start);
+		rodata_cap =
```

Comment:
This will no longer be NULL if there in the `!have_rodata_segment && !have_data_segment` case since it gets set to text_start?

Reply:
Misread the indentation lvel...